File: net.akehurst.language/parser/src/main/java/net/akehurst/language/grammar/parser/runtime/RuntimeRule.java
/**
 * Copyright (C) 2015 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.akehurst.language.grammar.parser.runtime;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

public class RuntimeRule {

	private final RuntimeRuleSet runtimeRuleSet;
	private final String name;
	private final int ruleNumber;
	private final RuntimeRuleKind kind;
	private final int patternFlags;
	private boolean isSkipRule;
	private RuntimeRuleItem rhs;
	private String toString_cache;

	RuntimeRule(final RuntimeRuleSet runtimeRuleSet, final String name, final int ruleNumber, final RuntimeRuleKind kind, final int patternFlags) {
		this.runtimeRuleSet = runtimeRuleSet;
		this.name = name;
		this.ruleNumber = ruleNumber;
		this.kind = kind;
		this.patternFlags = patternFlags;
	}

	public RuntimeRuleSet getRuntimeRuleSet() {
		return this.runtimeRuleSet;
	}

	public String getName() {
		return this.name;
	}

	public int getRuleNumber() {
		return this.ruleNumber;
	}

	public RuntimeRuleKind getKind() {
		return this.kind;
	}

	public int getPatternFlags() {
		return this.patternFlags;
	}

	// boolean isEmptyRule;

	public boolean getIsEmptyRule() {
		return null != this.getRhs() && this.getRhs().getKind() == RuntimeRuleItemKind.EMPTY;
	}

	public boolean getIsSkipRule() {
		return this.isSkipRule;
	}

	public void setIsSkipRule(final boolean value) {
		this.isSkipRule = value;
	}

	public void setRhs(final RuntimeRuleItem value) {
		this.rhs = value;
	}

	public RuntimeRuleItem getRhs() {
		return this.rhs;
	}

	public RuntimeRule getRhsItem(final int index) {
		if (this.getKind() == RuntimeRuleKind.TERMINAL) {
			return null;
		} else {
			return this.getRhs().getItems()[index];
		}
	}

	public int getRhsIndexOf(final RuntimeRule rule) {
		if (this.getKind() == RuntimeRuleKind.TERMINAL) {
			return -1;
		} else {
			return Arrays.asList(this.getRhs().getItems()).indexOf(rule);
		}
	}

	public RuntimeRule getSeparator() {
		return this.getRhsItem(1);
	}

	public RuntimeRule getRuleThatIsEmpty() {
		if (this.getIsEmptyRule()) {
			return this.getRhs().getItems()[0];
		} else {
			return null;
		}
	}

	Set<RuntimeRule> findAllNonTerminal() {
		final Set<RuntimeRule> result = new HashSet<>();
		if (RuntimeRuleKind.TERMINAL == this.getKind()) {
			return result;
		}
		for (final RuntimeRule item : this.getRhs().getItems()) {
			if (RuntimeRuleKind.NON_TERMINAL == item.getKind()) {
				result.add(item);
			}
		}
		return result;
	}

	Set<RuntimeRule> findAllNonTerminalAt(final int n) {
		final Set<RuntimeRule> result = new HashSet<>();
		if (RuntimeRuleKind.TERMINAL == this.getKind()) {
			return result;
		}

		final RuntimeRule[] items = this.getRhs().getItemAt(n);
		for (final RuntimeRule item : items) {
			if (RuntimeRuleKind.NON_TERMINAL == item.getKind()) {
				result.add(item);
			}
		}
		return result;
	}

	Set<RuntimeRule> findAllTerminal() {
		final Set<RuntimeRule> result = new HashSet<>();
		if (RuntimeRuleKind.TERMINAL == this.getKind()) {
			return result;
		}
		for (final RuntimeRule item : this.getRhs().getItems()) {
			if (RuntimeRuleKind.TERMINAL == item.getKind()) {
				result.add(item);
			}
		}
		switch (this.getRhs().getKind()) {
			case EMPTY:
			break;
			case CHOICE:
			break;
			case PRIORITY_CHOICE:
			break;
			case CONCATENATION:
			break;
			case MULTI: {
				if (0 == this.getRhs().getMultiMin()) {
					result.add(this.runtimeRuleSet.getEmptyRule(this));
				}
			}
			break;
			case SEPARATED_LIST: {
				if (0 == this.getRhs().getMultiMin()) {
					result.add(this.runtimeRuleSet.getEmptyRule(this));
				}
			}
			break;
			default:
			break;

		}
		return result;
	}

	/**
	 * depending on the kind of rule, determine if the given possibleChild rule could be a child at the given position
	 *
	 * @param position
	 * @return
	 */
	public boolean couldHaveChild(final RuntimeRule possibleChild, final int atPosition) {
		if (this.getKind() == RuntimeRuleKind.TERMINAL) {
			return false;
		} else {
			if (possibleChild.getIsSkipRule()) {
				return true;
			} else {
				switch (this.getRhs().getKind()) {
					case EMPTY:
						return false;
					case CHOICE:
						// TODO: cache this
						return 0 == atPosition && Arrays.asList(this.getRhs().getItems()).contains(possibleChild);
					case PRIORITY_CHOICE:
						return 0 == atPosition && Arrays.asList(this.getRhs().getItems()).contains(possibleChild);
					case CONCATENATION:
						return -1 == atPosition || atPosition >= this.getRhs().getItems().length ? false
								: this.getRhsItem(atPosition).getRuleNumber() == possibleChild.getRuleNumber();
					case MULTI:
						return this.getRhsItem(0).getRuleNumber() == possibleChild.getRuleNumber() || this.getRhs().getMultiMin() == 0 && atPosition == 0
								&& possibleChild.getIsEmptyRule() && possibleChild.getRuleThatIsEmpty().getRuleNumber() == this.getRuleNumber();
					case SEPARATED_LIST: {
						if (possibleChild.getIsEmptyRule()) {
							return this.getRhs().getMultiMin() == 0 && possibleChild.getRuleThatIsEmpty().getRuleNumber() == this.getRuleNumber();
						} else {
							if (atPosition % 2 == 0) {
								return this.getRhsItem(0).getRuleNumber() == possibleChild.getRuleNumber();
							} else {
								return this.getSeparator().getRuleNumber() == possibleChild.getRuleNumber();
							}
						}
					}
					default:
						throw new RuntimeException("Internal Error: rule kind not recognised " + this.getRhs().getKind());
				}
			}
		}
	}

	Set<RuntimeRule> findSubRules() {
		final Set<RuntimeRule> result = this.findAllNonTerminal();
		Set<RuntimeRule> oldResult = new HashSet<>();
		while (!oldResult.containsAll(result)) {
			oldResult = new HashSet<>();
			oldResult.addAll(result);
			for (final RuntimeRule nt : oldResult) {
				final Set<RuntimeRule> newNts = nt.findAllNonTerminal();
				newNts.removeAll(result);
				result.addAll(newNts);
			}
		}
		return result;
	}

	Set<RuntimeRule> findSubRulesAt(final int n) {
		final Set<RuntimeRule> result = this.findAllNonTerminalAt(n);
		Set<RuntimeRule> oldResult = new HashSet<>();
		while (!oldResult.containsAll(result)) {
			oldResult = new HashSet<>();
			oldResult.addAll(result);
			for (final RuntimeRule nt : oldResult) {
				final Set<RuntimeRule> newNts = nt.findAllNonTerminalAt(n);
				newNts.removeAll(result);
				result.addAll(newNts);
			}
		}
		return result;
	}

	public int incrementNextItemIndex(final int currentIndex) {
		switch (this.getRhs().getKind()) {
			case EMPTY:
				return -1;

			case CHOICE:
				return -1;

			case PRIORITY_CHOICE:
				return -1;

			case CONCATENATION:
				return this.getRhs().getItems().length == currentIndex + 1 ? -1 : currentIndex + 1;

			case MULTI:
				return currentIndex;

			case SEPARATED_LIST:
				return currentIndex == 0 ? 1 : 0;

			default:
				throw new RuntimeException("Internal Error: Unknown RuleKind " + this.getRhs().getKind());
		}
	}

	public Set<RuntimeRule> getNextExpectedItems(final int nextItemIndex, final int currentNumChildren) {
		return this.getRuntimeRuleSet().getNextExpectedItems(this, nextItemIndex, currentNumChildren);
	}

	public Set<RuntimeRule> findTerminalAt(final int n) {
		final Set<RuntimeRule> result = new HashSet<>();
		if (RuntimeRuleKind.TERMINAL == this.getKind()) {
			result.add(this);
			return result;
		}
		final RuntimeRule[] firstItems = this.getRhs().getItemAt(n);
		for (final RuntimeRule item : firstItems) {
			if (RuntimeRuleKind.TERMINAL == item.getKind()) {
				result.add(item);
			}
		}
		if (0 == n) {
			switch (this.getRhs().getKind()) {
				case EMPTY:
				break;
				case PRIORITY_CHOICE:
				break;
				case CHOICE:
				break;
				case CONCATENATION:
				break;
				case MULTI: {
					if (0 == n && 0 == this.getRhs().getMultiMin()) {
						result.add(this.runtimeRuleSet.getEmptyRule(this));
					}
				}
				break;
				case SEPARATED_LIST: {
					if (0 == n && 0 == this.getRhs().getMultiMin()) {
						result.add(this.runtimeRuleSet.getEmptyRule(this));
					}
				}
				break;
				default:
				break;

			}
		}
		return result;
	}

	public String getNodeTypeName() {
		return this.getName();
	}

	public String getTerminalPatternText() {
		return this.getName();
	}

	@Override
	public String toString() {
		if (null == this.toString_cache) {
			if (RuntimeRuleKind.NON_TERMINAL == this.getKind()) {
				this.toString_cache = "[" + this.getRuleNumber() + "](" + this.getNodeTypeName() + ") : " + this.getRhs();
			} else {
				this.toString_cache = "[" + this.getRuleNumber() + "]" + this.getTerminalPatternText();
			}
		}
		return this.toString_cache;
	}

	@Override
	public int hashCode() {
		return this.getRuleNumber();
	}

	@Override
	public boolean equals(final Object arg) {
		if (!(arg instanceof RuntimeRule)) {
			return false;
		}
		final RuntimeRule other = (RuntimeRule) arg;
		return this.getRuleNumber() == other.getRuleNumber();
	}

}

File: net.akehurst.language/agl-processor/src/commonMain/kotlin/agl/runtime/structure/RuntimeRule.kt
/**
 * Copyright (C) 2018 Dr. David H. Akehurst (http://dr.david.h.akehurst.net)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.akehurst.language.agl.runtime.structure

import net.akehurst.language.agl.api.runtime.Rule

/**
 * identified by: (runtimeRuleSetNumber, number, optionIndex)
 */
class RuntimeRule(
    val runtimeRuleSetNumber: Int,
    val ruleNumber: Int,
    val name: String?,
    override val isSkip: Boolean,
    override val isPseudo: Boolean
) : Rule {

    private lateinit var _rhs: RuntimeRuleRhs
    internal fun setRhs(value: RuntimeRuleRhs) {
        this._rhs = value
    }

    internal val rhs get() = this._rhs

    //TODO: needs properties- maybe:
    // isUnnamedLiteral - so we can eliminate from AsmSimple
    // isGenerated - also w.r.t. AsmSimple so we know if we should try and get a property name from the elements
    // not sure if I really want to add the data to this class as only used for AsmSimple not runtime use?

    val isExplicitlyNamed: Boolean get() = this.name != null
    override val tag: String get() = this.name ?: if (this.isTerminal) this.rhs.toString() else error("Internal Error: no tag")

    val isGoal get() = this.rhs is RuntimeRuleRhsGoal
    override val isEmptyTerminal get() = this.rhs is RuntimeRuleRhsEmpty
    override val isEmptyListTerminal get() = this.rhs is RuntimeRuleRhsEmptyList
    override val isEmbedded get() = this.rhs is RuntimeRuleRhsEmbedded
    override val isPattern get() = this.rhs is RuntimeRuleRhsPattern
    override val isLiteral get() = this.rhs is RuntimeRuleRhsLiteral

    /**
     * Empty, Literal, Pattern, Embedded
     */
    override val isTerminal
        get() = when (this.rhs) {
            is RuntimeRuleRhsNonTerminal -> false
            is RuntimeRuleRhsEmpty -> true
            is RuntimeRuleRhsEmptyList -> true
            is RuntimeRuleRhsLiteral -> true
            is RuntimeRuleRhsPattern -> true
            is RuntimeRuleRhsEmbedded -> true
            is RuntimeRuleRhsCommonTerminal -> true
        }

    /**
     * Goal, Concatenation, ListSimple, ListSeparated
     */
    val isNonTerminal
        get() = when (this.rhs) {
            is RuntimeRuleRhsTerminal -> false
            is RuntimeRuleRhsGoal -> true
            is RuntimeRuleRhsConcatenation -> true
            is RuntimeRuleRhsChoice -> true
            is RuntimeRuleRhsOptional -> true
            is RuntimeRuleRhsList -> true
        }

    override val isChoice get() = this.rhs is RuntimeRuleRhsChoice
    val isChoiceAmbiguous get() = this.isChoice && (this.rhs as RuntimeRuleRhsChoice).choiceKind == RuntimeRuleChoiceKind.AMBIGUOUS
    override val isOptional: Boolean get() = this.rhs is RuntimeRuleRhsOptional
    override val isListOptional: Boolean get() = this.rhs is RuntimeRuleRhsList && 0 == (this.rhs as RuntimeRuleRhsList).min
    override val isList get() = this.rhs is RuntimeRuleRhsList
    override val isListSimple: Boolean get() = this.rhs is RuntimeRuleRhsListSimple
    override val isListSeparated: Boolean get() = this.rhs is RuntimeRuleRhsListSeparated

    @Deprecated("use 'rhs is'")
//    val kind
//        get() = when {
//            isEmbedded -> RuntimeRuleKind.EMBEDDED
//            isGoal -> RuntimeRuleKind.GOAL
//            isTerminal -> RuntimeRuleKind.TERMINAL
//            isNonTerminal -> RuntimeRuleKind.NON_TERMINAL
//            else -> error("Internal Error")
//        }

    //val ruleThatIsEmpty: RuntimeRule get() = (this.rhs as RuntimeRuleRhsEmpty).ruleThatIsEmpty

    internal val asTerminalRulePosition by lazy { RulePosition(this, 0, RulePosition.END_OF_RULE) }

    //used in automaton build
    internal val rulePositions: Set<RulePosition> get() = rulePositionsAtStart + rulePositionsNotAtStart

    internal val rulePositionsNotAtStart: Set<RulePosition> get() = rhs.rulePositionsNotAtStart

    internal val rulePositionsAtStart get() = rhs.rulePositionsAtStart

    override val rhsItems get() = this.rhs.rhsItems

    override val unescapedTerminalValue: String
        get() = when (rhs) {
            is RuntimeRuleRhsLiteral -> (rhs as RuntimeRuleRhsLiteral).literalUnescaped
            is RuntimeRuleRhsPattern -> (rhs as RuntimeRuleRhsPattern).patternUnescaped
            else -> error("'unescapedTerminalValue' is only valid for Literals and Patterns")
        }

    //val rhsItems get() = this.rulePositions.flatMap { it.items }.toSet()

    val asString: String
        get() = when {
            isTerminal -> if (tag == rhs.asString) tag else "$tag(${rhs.asString})"
            isNonTerminal -> "$tag = ${rhs.asString}"
            else -> error("All rules should be either Terminal or NonTerminal")
        }

    // --- Any ---
    private val _hashCode = arrayOf(this.runtimeRuleSetNumber, this.ruleNumber).contentHashCode()
    override fun hashCode(): Int = _hashCode
    override fun equals(other: Any?): Boolean = when {
        other !is RuntimeRule -> false
        this.runtimeRuleSetNumber != other.runtimeRuleSetNumber -> false
        this.ruleNumber != other.ruleNumber -> false
        else -> true
    }

    override fun toString(): String = when {
        isTerminal -> if (tag == rhs.toString()) tag else "$tag($rhs)"
        isNonTerminal -> "$tag = $rhs"
        else -> error("All rules should be either Terminal or NonTerminal")
    }
}
